### ISSUE TITLE
library import problems: dereference symlinks broken?

### ISSUE THREAD

**Issue by bernt-matthias** [galaxyproject/galaxy#8478]
Trying to import from a user import directory using symlinks to the file system such that Galaxy stores the path to the actual file (ie the dereferenced symlink). But all options that I tried either store the path to the symlink or the file is copied. 

I have the following files, directories, and symlinks in the import directory

```
- test-file
- test-link-to-file
- test-link-to-dir
- testdir
- testdir/test-file
- testdir/test-link-to-file
- testdir/test-link-to-dir
- testdir2/test-link-to-file
```

In the import dialog this looks like this (more or less the same if `Choose Files` or `Choose Folders` is selected)

![Screenshot from 2019-08-19 14-58-23](https://user-images.githubusercontent.com/25689525/63267272-f223f500-c291-11e9-9139-5dc9cd64cd26.png)

First of all, independent of choosing `Choose Files` or `Choose Folders` symlinks to files are not shown (i.e., none of the test-link-to-file links is visible). The only difference between the two option seems to be that with `Choose Files` folders can't be selected (which might be reasonable), but links to folders can be selected in both cases (which might also be reasonable). Note that in both cases the heading of the dialog is `Please select folders or files`.

So there is no way to choose a single symlink to import. 

In case of `Choose Files`

- if I select the `test-link-to-dir` the import progress bar shows no progess (independent of `Link files instead of copying`)
- if a select a file in the `test-link-to-dir` and choose `Link files instead of copying` the path to the import dir stored. 

 `Choose Folders`

- If choosing `test-dir` with `Link files instead of copying` I also get a progress bar wo progress.
- If I choose `testdir2` with `Link files instead of copying` the contained link is not dereferenced.

I hope that I did not miss the option to actually do what I want to do. 

Bottom line is symlinks are not dereferenced and the import dialog is quite confusing.

**Comment by martenson** [comment]
what is your configuration for `user_library_import_symlink_whitelist` ?

**Comment by bernt-matthias** [comment]
I set `user_library_import_symlink_whitelist: "/DI/R1/,/DI/R2/,/DI/R3/"`

Where `/DI/R1/,/DI/R2/,/DI/R3/` are the only directories where users can have data.

**Comment by bernt-matthias** [comment]
Ahh. Maybe symlinks on our file system are the problem causing symlinks not to work here: 

Actually we have something like:

- `/home/` -> `/partition0/home/`
- `/data/` -> `/partition0/data/`
- `/work/` -> `/partition1/work/`

Currently the config variable lists only actual paths, i.e. `/partition...`

Have to check tomorrow if the symlinks in the library directory point to `/partition/X` or to `X`. I guess its `X`.

**Comment by martenson** [comment]
Briefly looking at the code the symlinks should resolve exactly once.

**Comment by bernt-matthias** [comment]
This would be exactly what I want. Lets see, I'm optimistic.

**Comment by bernt-matthias** [comment]
Does not work.

I have now `user_library_import_symlink_whitelist: "/home/,/partition0/home/,/data/,/partition0/data/,/work/,/partition1/work/"`

And tried to create the link in the import directory with both directories, ie. 

`ln -s /home/..../file test-link-to-file1` and `ln -s /partition0/home/.../test-link-to-file2`

I imported via `Chose Folder` and checked `Link files instead of copying`, since I can't chose the links with `Choose Files` as described above.

**Comment by martenson** [comment]
I spent some time today on this and found out that the API endpoint our UI uses does not support symlinks. There is another one, that is used e.g. by bioblend, that should work fine and dereference symlinks exactly once.

I will attempt to unify these endpoints under one codepath while making sure symlinks are understood by the one that the UI uses.

In the meantime if I were you I'd use bioblend to achieve what you need.

**Comment by bernt-matthias** [comment]
Good news

**Comment by bernt-matthias** [comment]
@martenson could you give me some information on the API endpoints you were writing about? Maybe I (or somebody else :)) could find time for this.
